### ISSUE TITLE
OpenWRT snapclient on GL-MT300A build incompatible to other snapcasts

### ISSUE THREAD

**Issue by vansoest** [badaix/snapcast#367]
By compiling snapclient for the latest stable OpenWRT/LEDE build in version v17.01.4 (git tag) on my GL-MT300A I got following errors:

## Snapclient on RAMIPS MT7620A GL-MT300A
```
# snapclient
2018-03-13 00-23-11 [Info] (Browser) NEW: service 'Snapcast #2' of type '_snapcast._tcp' in domain 'local'
2018-03-13 00-23-11 [Info] (Browser) NEW: service 'Snapcast' of type '_snapcast._tcp' in domain 'local'
2018-03-13 00-23-11 [Info] (Browser) CACHE_EXHAUSTED
2018-03-13 00-23-11 [Info] Service 'Snapcast #2' of type '_snapcast._tcp' in domain 'local':
2018-03-13 00-23-11 [Info] 	morphberry-2.local:5002 (192.168.1.3)
2018-03-13 00-23-11 [Info] Found server 192.168.1.3:5002
2018-03-13 00-23-11 [Info] Latency: 0
2018-03-13 00-23-11 [Notice] Connected to 192.168.1.3
2018-03-13 00-23-11 [Info] My MAC: "e4:95:6e:41:41:4d", socket: 5
2018-03-13 00-23-11 [Err] Controller::onException: std::bad_alloc
2018-03-13 00-23-11 [Err] Exception in Controller::worker(): std::bad_alloc
2018-03-13 00-23-11 [Err] Error in socket shutdown: Socket not connected
2018-03-13 00-23-12 [Notice] Connected to 192.168.1.3
2018-03-13 00-23-12 [Info] My MAC: "e4:95:6e:41:41:4d", socket: 5
2018-03-13 00-23-12 [Err] Controller::onException: std::bad_alloc
2018-03-13 00-23-12 [Err] Exception in Controller::worker(): std::bad_alloc
2018-03-13 00-23-13 [Notice] Connected to 192.168.1.3
2018-03-13 00-23-13 [Info] My MAC: "e4:95:6e:41:41:4d", socket: 5
2018-03-13 00-23-13 [Err] Controller::onException: std::bad_alloc
2018-03-13 00-23-13 [Err] Exception in Controller::worker(): std::bad_alloc
2018-03-13 00-23-13 [Err] Error in socket shutdown: Socket not connected
```

## Snapserver Arch Linux on the Raspberry Pi 2
```
2018-03-15 00-20-41 [Err] Error in socket shutdown: Transport endpoint is not connected
StreamServer::NewConnection: ::ffff:192.168.1.172
2018-03-15 00-20-42 [Notice] StreamServer::NewConnection: ::ffff:192.168.1.172
Exception in StreamSession::reader(): unknown message type received: 1280, size: 536936448
2018-03-15 00-20-42 [Err] Exception in StreamSession::reader(): unknown message type received: 1280, size: 536936448
2018-03-15 00-20-42 [Info] onDisconnect:
2018-03-15 00-20-42 [Err] Error in socket shutdown: Transport endpoint is not connected
StreamServer::NewConnection: ::ffff:192.168.1.172
2018-03-15 00-20-43 [Notice] StreamServer::NewConnection: ::ffff:192.168.1.172
Exception in StreamSession::reader(): unknown message type received: 1280, size: 536936448
2018-03-15 00-20-43 [Err] Exception in StreamSession::reader(): unknown message type received: 1280, size: 536936448
2018-03-15 00-20-43 [Info] onDisconnect:
2018-03-15 00-20-43 [Err] Error in socket shutdown: Transport endpoint is not connected
StreamServer::NewConnection: ::ffff:192.168.1.172
2018-03-15 00-20-44 [Notice] StreamServer::NewConnection: ::ffff:192.168.1.172
Exception in StreamSession::reader(): unknown message type received: 1280, size: 536936448
2018-03-15 00-20-44 [Err] Exception in StreamSession::reader(): unknown message type received: 1280, size: 536936448
2018-03-15 00-20-44 [Info] onDisconnect:
```

When I execute also the snapserver on the same machine (GL-MT300A) the `cat /dev/urandom > /tmp/snapfifo` test runs without problems!

## My steps to build the snapclient for the GL-MT300A router
used the docker image **bfoote/lede** and executed inside the container shell:
```git checkout v17.01.4
./scripts/feeds update -a
./scripts/feeds install -a
git clone https://github.com/badaix/snapos/
cp -Rv snapos/openwrt/ package/snapos
make defconfig
make menuconfig
make -j9
```

## Versions
* snapclient v0.13.0
* snapserver v0.13.0

Additionally, older snapclient versions currently with an older lede build v17.01.1 on another GL-MT300A runs without problems.

**Comment by badaix** [comment]
This looks like an endian issue. Can you please check if you have a `CMakeSystem.cmake` file in your build tree?

The line `set(CMAKE_SYSTEM_PROCESSOR "mips")` will cause CMake to add the `DIS_BIG_ENDIAN` preprocessor flag.


```
johannes@T400 ~/Develop/openwrt $ cat build_dir/target-mips_24kc_musl/snapcast-0.13.0/CMakeFiles/3.10.1/CMakeSystem.cmake 
set(CMAKE_HOST_SYSTEM "Linux-4.13.0-36-generic")
set(CMAKE_HOST_SYSTEM_NAME "Linux")
set(CMAKE_HOST_SYSTEM_VERSION "4.13.0-36-generic")
set(CMAKE_HOST_SYSTEM_PROCESSOR "x86_64")



set(CMAKE_SYSTEM "Linux-1")
set(CMAKE_SYSTEM_NAME "Linux")
set(CMAKE_SYSTEM_VERSION "1")
set(CMAKE_SYSTEM_PROCESSOR "mips")

set(CMAKE_CROSSCOMPILING "TRUE")

set(CMAKE_SYSTEM_LOADED 1)
```

**Comment by vansoest** [comment]
```
lede@lede-build:/opt/lede/build_dir/target-mipsel_24kc_musl-1.1.16/snapcast-0.13.0$ grep "SYSTEM" CMakeCache.txt
CMAKE_SYSTEM_NAME:UNINITIALIZED=Linux
CMAKE_SYSTEM_PROCESSOR:UNINITIALIZED=mipsel
CMAKE_SYSTEM_VERSION:UNINITIALIZED=1
```

removed the line DIS_BIG_ENDIAN for mips for testing:
https://github.com/vansoest/snapcast/commit/d57f2ec435faec2a4cdfb1994a1736b9c74424e3

but I think the condition isn't triggered because of it's defined as `UNINITIALIZED=mipsel` same problem.

**Comment by badaix** [comment]
I did some testing: your system is in fact declared as `mipsel`, which seems to match `if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "mips"` and thus the `IS_BIG_ENDIAN` flag is set in the CMakeListst.txt, although is it little endian.
In Snapcast's `CMakeList.txt` this block:
```
if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "mips"
    OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "armeb"
    OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "powerpc")
    set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -latomic")
    add_definitions("-DIS_BIG_ENDIAN")    
endif()
```
must be replaced with this block, which correctly identifies your CPU as little endian:
```
INCLUDE(TestBigEndian)
TEST_BIG_ENDIAN(BIGENDIAN)
IF(${BIGENDIAN})
    set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -latomic")
    add_definitions("-DIS_BIG_ENDIAN")    
ENDIF(${BIGENDIAN})
```

I will push the changes this evening in the develop branch.

**Comment by badaix** [comment]
Should be fixed now in the develop branch. Please try again with the develop branch of SnapOS.

**Comment by vansoest** [comment]
It works!

But my first attempt to compile the new code failed, because the atomic flag is for a 32bit CPU required (emulation) otherwise it can't compile like this snippet.:
```
timeProvider.cpp:(.text+0x5ac): undefined reference to `__atomic_store_8'
timeProvider.cpp:(.text+0x8ac): undefined reference to `__atomic_store_8'
```
The fix:
https://github.com/vansoest/snapcast/commit/ec02517ac4bd79e62b3c9e64f3df533f848386c0

for better performance to avoid xruns. I set the nice value to -19.:
```
snapclient -d'-19' $SNAPCLIENT_OPTS
```

**Comment by badaix** [comment]
fixed in release [v0.14.0](https://github.com/badaix/snapcast/releases/tag/v0.14.0)
